xds/src/main/java/io/grpc/xds/XdsDependencyManager.java

```diff
@@ -518,11 +518,21 @@ public void onChanged(XdsClusterResource.CdsUpdate update) {
 
             Set<String> addedClusters = Sets.difference(newNames, oldNames);
             Set<String> deletedClusters = Sets.difference(oldNames, newNames);
-            addedClusters.forEach((cluster) -> addWatcher(new CdsWatcher(cluster, parentContext)));
+            int depth = parentContexts.values().stream().max(Integer::compare).orElse(0) + 1;
+            if (depth > MAX_CLUSTER_RECURSION_DEPTH) {
+              logger.log(XdsLogger.XdsLogLevel.WARNING,
+                  "Cluster recursion depth limit exceeded for cluster {0}", resourceName());
+              Status error = Status.UNAVAILABLE.withDescription(
+                  "aggregate cluster graph exceeds max depth");
+              data = StatusOr.fromStatus(error);
+              throw error.asRuntimeException();
```

Comment:
Nothing catches this error, and nothing will see the `data` update, right? It seems this should not be throwing and instead we publish config that has failures.

Reply:
Updated.  Also found a number of other errors while dealing with this one.